### ISSUE TITLE
Set input rules to define allowed values for a given field

### ISSUE THREAD

**Issue by LouiseBonnard** [PrestaShop/PrestaShop#18400]
We currently have a lot of error notifications in the software stating that negative amounts are not allowed, or that users must type a positive value, etc. It is even worse for big pages, like the order page or the product page, that contain many fields, each expecting a specific value.

So when a user fills many fields, it becomes complicated if he made several mistakes since it displays several notifications. Yet, having too many error notifications is frustrating, over-complicates the page and confuses users in the end: what field? what value? where? what did I dooo?

😱

For a better experience, users should not be able to type whatever they want in fields but only the expected values. Something like the input is allowed only if it matches the required format.

Example: if the expected input is a quantity and a user types a letter (or any unwanted character) in this field, it should not work.

Of course, it might come with a help text right under the field but it would still be better: it will kind of 'localize' the solution (does not work, why, explanation is just below) and block the process right where it happens (no more fully filled pages ending up with a guessing game to find out what field to change).

_Originally posted by @LouiseBonnard in https://github.com/PrestaShop/PrestaShop/issues/18106#issuecomment-599583573_

**Comment by jolelievre** [comment]
Nice improvement! Indeed it's better for the user to have a feedback on his mistakes located on the incorrect input (via a tooltip or other).

And imho an error message is easier to understand than blocking an invalid input. For example if I type `-1` in an input and it automatically transforms it into `1` or `0` I won't understand why it doesn't type what I want.

But leaving the `-1` with an error color on the input and an help message helps me see and understand why it's an invalid value

**Comment by LouiseBonnard** [comment]
I like the idea of the color highlighting the error input!

**Comment by jolelievre** [comment]
Yes but it can't be the only way to display the error (for colorblind people)

**Comment by prestascott** [comment]
Hello everyone!

Thank you @LouiseBonnard for the request. 

I understand that it can be frustrating for users when they receive 4 error messages after filling up many inputs. 
But imo, we shouldn't block them by disabling letters when numbers are the only valid characters in the input. 
They might understand it's a bug from their keyboard or from PrestaShop but in reality, we did that in purpose.
What we want here is to inform users that they did something wrong by putting letters instead of numbers. 

I suggest to highlight the input when they start typing the wrong characters. Input will automatically display the error message that must be explicit.
![errorinput1](https://user-images.githubusercontent.com/54888056/99373413-aeb79c80-28c1-11eb-9232-c5e8547f5d9e.png)

Or, the highlight and the error message can appear when they finished filling up the input (for example, when they click on the next input to fill up). It allows user to move forward step by step correctly. 
![errorinput2](https://user-images.githubusercontent.com/54888056/99374264-b3308500-28c2-11eb-8e4b-c8cb3ac412cb.png)

With these 2 solutions, users will see where they did something wrong before clicking the "Save" button. 
They also avoid receiving many error messages at the same time.

Wdyt?

**Comment by LouiseBonnard** [comment]
Nice examples, thank you, @prestascott! I think both solutions will help users better understand what to type and that's great. I prefer the first one since it warns users _while they fill in the field_, I see it as a way of saving time. If we can have a dynamic checking of the information typed in the field to display the error message directly, I believe it's the best solution. :-)

**Comment by Hlavtox** [comment]
Already implemented, errors are nice.
